### PR TITLE
#30

### PR DESCRIPTION
**Pull request by Shubhamd13**

### CODE REVIEW

File: src/main/java/com/kodekonveyor/work_request/create/CreateWorkRequestController.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	private WorkRequestRepository workRequestRepository;
+
+	@PostMapping("/work-request")
+	public void call(@RequestBody final CreateWorkRequestDTO createWorkRequestDTO) {
+		final WorkRequestEntity workRequestEntity = new WorkRequestEntity();
+		workRequestEntity.setWorkType(createWorkRequestDTO.getWorkType());
+		final UserEntity userEntity = authenticatedUserService.call();
+		workRequestEntity.setCustomer(userEntity);
+		workRequestEntity.setId(createWorkRequestDTO.getCustomerId());
+
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/work_request/create/CreateWorkRequestController.java]: setDescription is missing. Please set all the fields coming from the DTO. Having a test for each of the fields is ok.

File: src/main/java/com/kodekonveyor/work_request/open/WorkRequestEntity.java [hunk 2]
```diff
@@ -10,4 +10,8 @@
 
 	private String workType;
 
+	public void save() {
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/work_request/open/WorkRequestEntity.java]: This save has no functionality. Please remove it.

Comment (szirbucz) [reply src/main/java/com/kodekonveyor/work_request/open/WorkRequestEntity.java]: Please take care about Entities and DTOs supposed to contain no methods.

File: src/main/java/com/kodekonveyor/work_request/open/WorkRequestEntity.java [hunk 3]
```diff
@@ -10,4 +10,8 @@
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/work_request/open/WorkRequestEntity.java]: The Entity has no ID yet. Please add the following field
```
@Id
@GeneratedValue(strategy = GenerationType.IDENTITY)
 private long id;
```

File: src/test/java/com/kodekonveyor/authentication/AuthenticatedUserStubs.java [hunk 4]
```diff
@@ -1,11 +1,11 @@
-
 package com.kodekonveyor.authentication;
 
 import static org.mockito.Mockito.doReturn;
 
-class AuthenticatedUserStubs {
+public class AuthenticatedUserStubs { // NOPMD
```
Comment (szirbucz) [inline src/test/java/com/kodekonveyor/authentication/AuthenticatedUserStubs.java]: It seems, you fixed the PMD error by adding "public". "// NOPMD" is not needed in this case.

File: src/test/java/com/kodekonveyor/work_request/WorkRequestTestData.java [hunk 5]
```diff
@@ -29,6 +31,9 @@ public WorkRequestTestData(final UserTestData userTestData) {
 		WORK_REQUEST_LIST_DTO = new WorkRequestListDTO();
 		WORK_REQUEST_LIST_DTO.setRequests(List.of(WORK_REQUEST_DTO));
 		WORK_REQUEST_ENTITY_LIST = List.of(WORK_REQUEST_ENTITY);
+		CREATE_WORK_REQUEST = new CreateWorkRequestDTO();
+		CREATE_WORK_REQUEST.setCustomerId(WORK_REQUEST_ID);
+		CREATE_WORK_REQUEST.setWorkType(WORK_TYPE);
```
Comment (szirbucz) [inline src/test/java/com/kodekonveyor/work_request/WorkRequestTestData.java]: Please refactor these 3 lines to a create method.
```
private void createCreateWorkRequest() {
    CREATE_WORK_REQUEST = new CreateWorkRequestDTO();
    CREATE_WORK_REQUEST.setCustomerId(WORK_REQUEST_ID);
    CREATE_WORK_REQUEST.setWorkType(WORK_TYPE);
}
```

See TestData 7c https://docs.google.com/document/d/1zKO8gJ8rdPUUkj6xiti_E08Y4QKO2AB4F-qO5__Dtb0/edit?usp=sharing

File: src/test/java/com/kodekonveyor/work_request/create/CreateWorkRequestControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+		AuthenticatedUserStubs.behaviour(authenticatedUserService, userTestData);
+	}
+
+	@Test
+	@DisplayName("Controller creates WorkRequestEntity based on request")
+	public void test() {
+		createWorkRequestController.call(workRequestTestData.CREATE_WORK_REQUEST);
+		verify(workRequestRepository).save(workRequestTestData.WORK_REQUEST_ENTITY);
+	}
+
```
Comment (szirbucz) [inline src/test/java/com/kodekonveyor/work_request/create/CreateWorkRequestControllerTest.java]: Please create test cases for each fields.

File: src/test/java/com/kodekonveyor/work_request/WorkRequestTestData.java (hunk 5 above)
Comment (Shubhamd13) [reply src/test/java/com/kodekonveyor/work_request/WorkRequestTestData.java]: CREATE_WORK_REQUEST requires AddressTestData.

Comment (Shubhamd13) [reply src/test/java/com/kodekonveyor/work_request/WorkRequestTestData.java]: How would i pass AddressTestData to a non static private method for initializing a final field.

Comment (Shubhamd13) [reply src/test/java/com/kodekonveyor/work_request/WorkRequestTestData.java]: @szirbucz

Comment (Shubhamd13) [reply src/test/java/com/kodekonveyor/work_request/WorkRequestTestData.java]: Made the necessary changes after a call with Árpád.

File: src/test/java/com/kodekonveyor/work_request/WorkRequestTestData.java [hunk 7]
```diff
@@ -16,19 +17,47 @@
...
 		this.userTestData = userTestData;
-		WORK_REQUEST_ENTITY = new WorkRequestEntity();
-		WORK_REQUEST_ENTITY.setCustomer(userTestData.USER);
-		WORK_REQUEST_ENTITY.setWorkType(WORK_TYPE);
-		WORK_REQUEST_ENTITY.setId(WORK_REQUEST_ID);
-		WORK_REQUEST_DTO = new WorkRequestDTO();
-		WORK_REQUEST_DTO.setWorkRequestId(WORK_REQUEST_ID);
-		WORK_REQUEST_DTO.setWorkType(WORK_TYPE);
+		this.addressTestData = addressTestData;
+		ADDRESS_ENTITY = this.addressTestData.ADDRESS_ENTITY;
```
Comment (magwas) [inline src/test/java/com/kodekonveyor/work_request/WorkRequestTestData.java]: We do not need ADDRESS_ENTITY here as it is already in AddressTestData. Please remove this line and the field.

File: src/test/java/com/kodekonveyor/work_request/WorkRequestTestData.java [hunk 8]
```diff
@@ -16,19 +17,47 @@
...
+		ADDRESS_DTO = this.addressTestData.ADDRESS_DTO;
```
Comment (magwas) [inline src/test/java/com/kodekonveyor/work_request/WorkRequestTestData.java]: same as above

**Comment by szirbucz** [comment]
testcases=5

**Comment by szirbucz** [comment]
Great job @Shubhamd13 . The last step before you get your compensation is filling the Developers Feedback form.
https://docs.google.com/forms/d/e/1FAIpQLSdNMbUDPHp2CCbsKtx3qMDCGXIYmmRsjJoF201Xtl-xvb2llQ/viewform

**Pull request merged**